tests/data/test_module.py

```diff
@@ -10,36 +11,83 @@
 
 @pytest.mark.skipif(GITHUB_ACTIONS, reason="only works locally")
 class TestModuleInstantiation:
-    def test_datamodule_initialization(self):
-        temp_data_args = ClimateDatasetArgs(variables=["2m_temperature"], split="train")
-        temp_highres_data_args = ClimateDatasetArgs(
-            variables=["2m_temperature"], split="train"
-        )
-        temp_task_args = DownscalingArgs(
-            temp_data_args,
-            temp_highres_data_args,
-            in_vars=["2m_temperature"],
-            out_vars=["2m_temperature"],
-        )
-        DataModuleArgs(temp_task_args, 2014, 2016, 2017)
-
-    def test_initialization(self):
-        temp_data_args = ERA5Args(
+    def test_map_initialization(self):
+        climate_dataset_args = ERA5Args(
             root_dir=os.path.join(DATA_PATH, "era5/5.625deg/"),
-            variables=["2m_temperature"],
-            years=range(2010, 2015),
+            variables=[
+                "2m_temperature",
+                "geopotential",
+                "temperature_250",
+                "orography",
+                "land_sea_mask",
+            ],
+            years=range(2009, 2015),
             split="train",
         )
-        temp_highres_data_args = ERA5Args(
+        high_res_climate_dataset_args = ERA5Args(
             root_dir=os.path.join(DATA_PATH, "era5/2.8125deg/"),
-            variables=["2m_temperature"],
-            years=range(2010, 2015),
+            variables=[
+                "geopotential",
+                "temperature_250",
+            ],
+            years=range(2009, 2015),
+            split="train",
+        )
+        stacked_climate_dataset_args = StackedClimateDatasetArgs(
+            [climate_dataset_args, high_res_climate_dataset_args]
+        )
+        downscaling_args = DownscalingArgs(
+            in_vars=[
+                "2m_temperature",
+                "geopotential",
+                "temperature_250",
+            ],
+            out_vars=[
+                "geopotential",
+                "temperature_250",
+            ],
+            constant_names=[
+                "orography",
+                "land_sea_mask",
+            ],
+            subsample=6,
+        )
+        dataset_args = ShardDatasetArgs(
+            stacked_climate_dataset_args, downscaling_args, 2
+        )
+
+        DataModule(DataModuleArgs(dataset_args, 2009, 2015, 2017))
```

Comment:
Can we change this so the user doesn't have to specify the years again here? They already specified the years as 2009-2015 in the `ERA5Args` instances. Also, IMO it isn't clear why 2017 is here. I know that's the start of the test years but shouldn't it be used as some point before we specify `DataModuleArgs`?

Reply:
I think you got confused with the 2009-2015 in the `ERA5Args`. That really doesn't do anything. The `DataModuleArgs` overwrite the `years` of `ERA5Args` when it calls the `ERA5Args`'s `setup()`. Based on the split, the logic in `ERA5Args`'s `setup()`just copies the start year and end year from `train_start_year`, `val_start_year`, `test_start_year` and `test_end_year`.

For more details refer to my comment on "why we need StackedClimateArgs" class.

Reply:
Sorry can you link the comment you are talking about? Not sure which one.

Reply:
Also, not a huge fan of the user specifying 2009-2015 in the `ERA5Args` and it not doing anything. Pretty sure it's bad practice to have arguments that don't have an effect.

Reply:
I thought more about it and I realized that yes, it is not a good practice. Thus I have decided to get rid of `DataModuleArgs`. Now the [`DataModule`](https://github.com/aditya-grover/climate-learn/blob/2f3db813b98ccb91b14709da7796734e5bdba99e/src/climate_learn/data/module.py#L65) takes `train_dataset_args`, `val_dataset_args`, `test_dataset_args` as arguments instead of `DataModuleArgs` where each of them should be an instance of either [`MapDatasetArgs`](https://github.com/aditya-grover/climate-learn/blob/2f3db813b98ccb91b14709da7796734e5bdba99e/src/climate_learn/data/dataset/args/map_dataset_args.py#L15) or [`ShardDatasetArgs`](https://github.com/aditya-grover/climate-learn/blob/2f3db813b98ccb91b14709da7796734e5bdba99e/src/climate_learn/data/dataset/args/shard_dataset_args.py#L15)

Now, since you pointed out that user may not want to manually create different instances of ClimateDatasetArgs class, TaskArgs class, etc. Since all of them would be pretty similar and the only difference being just the [`years`](https://github.com/aditya-grover/climate-learn/blob/2f3db813b98ccb91b14709da7796734e5bdba99e/src/climate_learn/data/climate_dataset/args/era5_args.py#L25) attribute and [`split`](https://github.com/aditya-grover/climate-learn/blob/2f3db813b98ccb91b14709da7796734e5bdba99e/src/climate_learn/data/climate_dataset/args/climate_dataset_args.py#L23) for `ERA5`. Thus, I have added new functions called [`create_copy()`](https://github.com/aditya-grover/climate-learn/blob/2f3db813b98ccb91b14709da7796734e5bdba99e/src/climate_learn/data/climate_dataset/args/climate_dataset_args.py#L26) and [`check_validity()`](https://github.com/aditya-grover/climate-learn/blob/2f3db813b98ccb91b14709da7796734e5bdba99e/src/climate_learn/data/climate_dataset/args/climate_dataset_args.py#L34). The first function takes as input a dict where the keys are the attribute name and the value is what you want to set it to. The second function is used to test whether the modified copy is a valid config.

See [`test_module.py`](https://github.com/aditya-grover/climate-learn/blob/2f3db813b98ccb91b14709da7796734e5bdba99e/tests/data/test_module.py#L109) for reference.